### PR TITLE
Zijuan

### PR DESCRIPTION
**Pull request by ZijuanZhang**
mainly special methods

### CODE REVIEW

**Review by JesseZhuang — CHANGES_REQUESTED** [review]
What did you modify in the file `numeric_math/Speed_up_computation_in_Python.ipynb`?

**Comment by ZijuanZhang** [comment]
nothing, I am also wondering why it is shows as I changed it. I tried to revoke the commit, but it seems much more messy, so I just PR as this, sorry about the confusion.

File: text_processing/RegExpPart1.py [hunk 1]
```diff
@@ -97,3 +97,11 @@
...
 print("Back reference group(1):", mo.group(1))
 print("Back reference group(2):", mo.group(2))
+
+print(re.search('[^aeiou]y$', 'vacancy'))
+# not raise error
+
+print(re.sub('([^aeiou])y$', r'\1ies', 'vacancy'))
+# ([^aeiou])y$ the word that ends with not aeiou + y
+# \1 group 1; group 0 is the whole content
+# raise error if there is no parenthesis in .sub method, () forms group?
```
Comment (JesseZhuang) [inline text_processing/RegExpPart1.py]: correct, () forms group. just as shown in lines above your added code 96-99. You can update this now.

File: special_method.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    def __str__(self):
+        return str(self.converse2meters())
+
+    def __repr__(self):
+        return "Length(" + str(self.value) + ", '" + self.unit + "')"
+
+if __name__ == "__main__":
+    x = Length(4)
+    print(x)
+    y = eval(repr(x))
```
Comment (JesseZhuang) [inline special_method.py]: `y` not used

File: text_processing/RegExpPart1.py (hunk 1 above)
Comment (JesseZhuang) [inline text_processing/RegExpPart1.py]: correct, () forms group. as shown in lines 96-99. you can update this.

File: special_method.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    z = Length(4.5, "yd") + Length(1)
+    print(repr(z))
+    print(z)
+
+    L = Length
```
Comment (JesseZhuang) [inline special_method.py:47]: what is this called? class aliasing?

**Review by JesseZhuang — CHANGES_REQUESTED** [review]
Next time before you do git commit please make sure there is no unwanted changes.

Definitely do not do git push if you have doubt on your last commit.

Try this to fix your wrong push.
http://stackoverflow.com/questions/1270514/undoing-a-git-push

**Comment by JesseZhuang** [comment]
It's your first push which introduced error: https://github.com/InCodeLearning/InCodeLearning-Python3/pull/44/commits/0472d12143b1d1ba63cb407800f947bae8b39da7

Also your commit messages did not describe what was done for that particular commit.

File: special_method.py (hunk 3 above)
Comment (ZijuanZhang) [reply special_method.py:47]: I think this can be called as class instance.

File: special_method.py (hunk 2 above)
Comment (ZijuanZhang) [reply special_method.py]: this can be deleted.

**Pull request closed without merging**